### PR TITLE
Group 4: For Loop

### PR DESCRIPTION
**Pull request by maxschwinghammer**
Test fails, but we don't know why and can't fix it

### CODE REVIEW

File: src/main/java/com/auberer/compilerdesignlectureproject/ast/ASTForNode.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        Set<TokenType> followSet = Set.of(TokenType.TOK_RBRACE, TokenType.TOK_DEFAULT, TokenType.TOK_RETURN);
+
+        Set<TokenType> selectionSet = new HashSet<>(ASTStmtNode.getSelectionSet());
+
+        Set<TokenType> forLoopSelectionSet = new HashSet<>();
+        forLoopSelectionSet.add(TokenType.TOK_FOR);
+        selectionSet.addAll(forLoopSelectionSet);
+
+        selectionSet.addAll(followSet);
+        return selectionSet;
```
Comment (marcauberer) [inline src/main/java/com/auberer/compilerdesignlectureproject/ast/ASTForNode.java:24]: The above code is not quite right. Remember, this is the rule for selection sets:
`SELECT(nonTerminal) = FIRST(nonTerminal)` if `FIRST(nonTerminal)` contains epsilon. Otherwise `SELECT(nonTerminal) = FIRST(nonTerminal) \ epsilon U FOLLOW(nonTerminal)`

File: src/test/java/com/auberer/compilerdesignlectureproject/reader/test.txt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+This is a test file.
```
Comment (marcauberer) [inline src/test/java/com/auberer/compilerdesignlectureproject/reader/test.txt:1]: Have you rebased onto upstream? I think that shouldn't be part of the PR.

File: src/test/java/com/auberer/compilerdesignlectureproject/parser/PrintBuiltinCallNodeTest.java [hunk 3]
```diff
@@ -30,7 +30,7 @@ void setUp() {
 
   @Test
   @DisplayName("Test print builtin call")
-  void testPrintBuiltinCall() {
+  void testPrintBuiltinCall() throws Exception {
```
Comment (marcauberer) [inline src/test/java/com/auberer/compilerdesignlectureproject/parser/PrintBuiltinCallNodeTest.java:33]: Why is that required?

File: src/main/java/com/auberer/compilerdesignlectureproject/parser/Parser.java [hunk 4]
```diff
@@ -42,6 +42,28 @@ public ASTEntryNode parse() {
     return node;
   }
 
+  public ASTForNode parseForLoop() throws Exception {
+    ASTForNode node = new ASTForNode();
+
+    enterNode(node);
+
+    Set<TokenType> forSelectionSet = ASTForNode.getSelectionSet();
```
Comment (marcauberer) [inline src/main/java/com/auberer/compilerdesignlectureproject/parser/Parser.java:50]: This should be unused?

File: src/main/java/com/auberer/compilerdesignlectureproject/parser/Parser.java [hunk 5]
```diff
@@ -42,6 +42,28 @@ public ASTEntryNode parse() {
...
+    lexer.expect(TokenType.TOK_FOR);
+    lexer.expect(TokenType.TOK_LPAREN);
+    parseAssignExpr();
+    lexer.expect(TokenType.TOK_SEMICOLON);
+    parseAssignExpr();
+    lexer.expect(TokenType.TOK_SEMICOLON);
+    parseAssignExpr();
+    lexer.expect(TokenType.TOK_RPAREN);
+    lexer.expect(TokenType.TOK_LBRACE);
+    //parseStmtLst();
```
Comment (marcauberer) [inline src/main/java/com/auberer/compilerdesignlectureproject/parser/Parser.java:61]: Why is this commented out?

File: src/test/java/com/auberer/compilerdesignlectureproject/parser/ForNodeTest.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        doNothing().when(lexer).expect(TokenType.TOK_FOR);
+        doNothing().when(lexer).expect(TokenType.TOK_LPAREN);
+        doNothing().when(parser).parseAssignExpr();
+        doNothing().when(lexer).expect(TokenType.TOK_SEMICOLON);
+        doNothing().when(parser).parseAssignExpr();
+        doNothing().when(lexer).expect(TokenType.TOK_SEMICOLON);
+        doNothing().when(parser).parseAssignExpr();
+        doNothing().when(lexer).expect(TokenType.TOK_RPAREN);
+        doNothing().when(lexer).expect(TokenType.TOK_LBRACE);
+        doReturn(mock(ASTStmtLstNode.class)).when(parser).parseStmtLst();
```
Comment (marcauberer) [inline src/test/java/com/auberer/compilerdesignlectureproject/parser/ForNodeTest.java:54]: Simply return null or something instead of the mock here.

File: src/test/java/com/auberer/compilerdesignlectureproject/parser/ForNodeTest.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        MockitoAnnotations.openMocks(this);
+        parser = new Parser(lexer);
+        parser = spy(parser);
+    }
+
+    // Test fails, but we don't know why and can't fix it
+
+    @Test
+    @DisplayName("Test for node")
+    void testForNode() throws Exception {
```
Comment (marcauberer) [inline src/test/java/com/auberer/compilerdesignlectureproject/parser/ForNodeTest.java:42]: Please catch all exceptions within the test body.

**Review by marcauberer — COMMENTED** [review]
Thanks! I noted a few things. Let's see if all tests pass after you fixed those. Otherwise, I will take a closer look.

**Comment by maxschwinghammer** [comment]
> Thanks! I noted a few things. Let's see if all tests pass after you fixed those. Otherwise, I will take a closer look.

We have responded to your comments and pushed new changes, but unfortunately the tests are not working

**Comment by marcauberer** [comment]
> > Thanks! I noted a few things. Let's see if all tests pass after you fixed those. Otherwise, I will take a closer look.
> 
> We have responded to your comments and pushed new changes, but unfortunately the tests are still not working

Will merge anyways and fix the tests in-branch. Thank you!
